### PR TITLE
kraken: Night bus filter linearisation

### PR DESCRIPTION
**Pull request by nonifier**
Associated PR:

- [x] https://github.com/CanalTP/navitia-proto/pull/115
- [x] https://github.com/CanalTP/utils/pull/66

### CODE REVIEW

File: source/ed/build_helper.h [hunk 1]
```diff
@@ -118,7 +118,7 @@ struct VJ {
 
     VJ& operator()(const std::string& stopPoint,
                    const std::string& arrivee,
-                   const std::string& depart,
+                   const std::string& depart = "",
```
Comment (mbounabi) [inline source/ed/build_helper.h]: French? arrivée, départ?

File: source/routing/journey.h [hunk 2]
```diff
@@ -83,5 +84,40 @@ struct SectionHash {
...
+
+    auto latest_journey = [](const Journey& j1, const Journey& j2) {
+        return j1.arrival_dt > j2.arrival_dt;
+    };
+
+    const auto best_element = clockwise ?
+                                std::min_element<Iterator, Compare_func> :
+                                std::max_element<Iterator, Compare_func>;
+    const auto comp_journeys = clockwise ? earliest_journey : latest_journey;
+    const auto best = best_element(journeys.cbegin(), journeys.cend(), comp_journeys);
```
Comment (xlqian) [inline source/routing/journey.h]: how about
```c++
const auto best = clockwise ? boost::min_element(journeys, earliest_journey) :  boost::max_element(journeys, latest_journey);
```
and remove all `typedef` ? no?

File: source/routing/tests/journey_test.cpp [hunk 3]
```diff
@@ -45,89 +45,85 @@ struct logger_initialized {
...
-        stop_times.push_back(vj3.stop_times[1].st);
+        j.sections.push_back(section);
 
-        return stop_times;
+        return j;
     }
-}
 
-BOOST_AUTO_TEST_CASE(similar_sections_should_be_equal) {
+    std::vector<type::StopTime>& make_stop_times() { return stop_times; }
```
Comment (xlqian) [inline source/routing/tests/journey_test.cpp]: not sure i understand this well,  is this function still useful?

File: source/ed/build_helper.h (hunk 1 above)
Comment (nonifier) [reply source/ed/build_helper.h]: Do you parlez vous ?!

File: source/routing/journey.h (hunk 2 above)
Comment (nonifier) [reply source/routing/journey.h]: nice one !

File: source/routing/tests/journey_test.cpp (hunk 3 above)
Comment (nonifier) [reply source/routing/tests/journey_test.cpp]: Nope, it isn't indeed ! it's now gone ;)

File: source/routing/journey.h [hunk 4]
```diff
@@ -83,5 +84,35 @@ struct SectionHash {
...
+ *
+ * @param journeys A container of journeys
+ * @param clokwise Active clockwise or not
+ * @return best jouney
+ */
+template<class Journeys>
+const Journey& get_best_journey(const Journeys & journeys, bool clockwise)
+{
+    if(journeys.size() == 0)
+        throw std::invalid_argument("get_best_journey takes a list of at least 1 journey");
```
Comment (kinnou02) [inline source/routing/journey.h]: use a ```navitia::recoverable_exception```

**Review by benoit-bst — APPROVED** [review]

**Review by kinnou02 — APPROVED** [review]

**Review by TeXitoi — APPROVED** [review]

**Pull request merged**